symfit/core/fit.py

```diff
@@ -1009,169 +1010,78 @@ class HasCovarianceMatrix(TakesData):
-        for data in self.dependent_data.values():
-            if flatten:
-                if data is not None:
-                    degrees_of_freedom += np.product(data.shape)
+        if isinstance(objective, LogLikelihood):
+            # Loglikelihood is a special case that needs to be considered
+            # separately, see #138.
+            return hess_inv
+        elif isinstance(objective, LeastSquares):
+            # Calculate the covariance for a least squares method.
+            # https://www8.cs.umu.se/kurser/5DA001/HT07/lectures/lsq-handouts.pdf
+            # Residual sum of squares
+            rss = objective(**key2str(best_fit_params))
+            # Degrees of freedom
+            raw_dof = np.sum([np.product(shape) for shape in self.data_shapes[1]])
+            dof = raw_dof - len(self.model.params)
+            if self.absolute_sigma:
+                # When interpreting as measurement error, we do not rescale.
+                s2 = 1
             else:
-                if data is not None:
-                    degrees_of_freedom.append(np.product(data.shape))
-                    # degrees_of_freedom = np.product(data.shape) - len(popt)
-                    # break
-                else: # the correspoding component in ss_res will be 0 so it is ok to add any non-zero number.
-                    degrees_of_freedom.append(len(best_fit_params) + 1)
-        degrees_of_freedom = np.array(degrees_of_freedom)
-        s_sq = ss_res / (degrees_of_freedom - len(best_fit_params))
-        # s_sq = ss_res / degrees_of_freedom
-
-        return s_sq
+                s2 = rss / dof
+            # Divide by two because the source uses different normalization
+            cov_mat = 2 * s2 * hess_inv
+            return cov_mat
+        else:
+            # We do not know how to handle with this situation.
+            return None
 
-    def _cov_mat_equal_lenghts(self, best_fit_params):
+    def covariance_matrix(self, best_fit_params):
         """
-        If all the data arrays are of equal size, use this method. This will
-        typically be the case, and this method is a lot faster because it allows
-        for numpy magic.
+        Given best fit parameters, this function finds the covariance matrix.
+        This matrix gives the (co)variance in the parameters.
 
         :param best_fit_params: ``dict`` of best fit parameters as given by .best_fit_params()
+        :return: covariance matrix.
         """
-        # Stack in a new dimension, and make this the first dim upon indexing.
-        sigma = np.concatenate([arr[np.newaxis, ...] for arr in self.sigma_data.values()], axis=0)
+        cov_matrix = self._covariance_matrix(best_fit_params,
+                                             objective=self.objective)
+        if cov_matrix is None:
+            # If the covariance matrix could not be computed, we try again by
+            # approximating the hessian with the jacobian.
```

Comment:
This bit of code would become cleaner if there were 2 methods: `_cov_matrix_from_hessian` (is now `_covariance_matrix`, and `_cov_matrix_from_jacobian`

Reply:
That would also help in testing: is _cov_matrix_from_hessian ~= _cov_matrix_from_jacobian?